Kernel/FileSystem/FIFO.cpp

```diff
@@ -60,6 +60,27 @@ NonnullRefPtr<FileDescription> FIFO::open_direction(FIFO::Direction direction)
     return description;
 }
 
+NonnullRefPtr<FileDescription> FIFO::open_direction_blocking(FIFO::Direction direction)
+{
+    auto description = open_direction(direction);
+
+    if (direction == Direction::Reader) {
+        m_read_open_queue.wake_all();
+
+        if (m_writers == 0)
+            Thread::current()->wait_on(m_write_open_queue, "FIFO");
+    }
+
+    if (direction == Direction::Writer) {
+        m_write_open_queue.wake_all();
+
+        if (m_readers == 0)
```

Comment:
Is there any locking here?

Reply:
Good catch, I now lock the fifo unless we are waiting for the other end to be opened.

Reply:
If you're referring to deadlock, then that won't happen because when one end is blocking it's `m_readers`/`m_writers` is already incremented, so the other side would never block.